src/core/PwAuth.cpp

```diff
@@ -0,0 +1,193 @@
+#include "PwAuth.hpp"
+#include "ext-session-lock-v1-protocol.h"
+#include "hyprlock.hpp"
+#include "../helpers/Log.hpp"
+#include "../config/ConfigManager.hpp"
+#include <cstddef>
+#include <cstring>
+#define GCRYPT_NO_DEPRECATED
+#define GCRYPT_NO_MPI_MACROS
+#define NEED_LIBGCRYPT_VERSION nullptr
+#include <gcrypt.h>
+
+using namespace std::chrono_literals;
+
+static auto hex2Bytes(const std::string& hex) -> std::unique_ptr<unsigned char[]> {
+    auto bytes = std::make_unique<unsigned char[]>(hex.length() / 2u);
+    for (std::size_t i = 0; i < hex.length(); i += 2) {
+        try {
+            auto v = std::stoi(hex.substr(i, 2), nullptr, 16);
+            if (v >= 0) {
+                bytes[i / 2] = static_cast<unsigned char>(v);
+            } else
+                throw std::invalid_argument("invalid hex value");
+        } catch (std::invalid_argument const& e) {
+            Debug::log(ERR, "auth: invalid password_hash");
+            bytes = nullptr;
+        } catch (std::out_of_range const& e) {
+            // Should never happen, as 2-byte substrings should never go o-o-r.
+            Debug::log(CRIT, "auth: implementation error in hex2Bytes conversion");
+            bytes = nullptr;
+        }
+    }
+    return bytes;
+}
+
+static auto bytes2Hex(const unsigned char* bytes, std::size_t len) -> std::string {
+    std::stringstream ss;
+    ss << std::setw(2) << std::setfill('0') << std::hex;
+    for (std::size_t i = 0; i < len; ++i)
+         ss << (int)bytes[i];
+    return ss.str();
+}
+
+CPwAuth::CPwAuth() {
+
+    if (gcry_check_version(NEED_LIBGCRYPT_VERSION))
+        gcry_control(GCRYCTL_INITIALIZATION_FINISHED, 0);
+    else
+        Debug::log(CRIT, "libgcrypt too old");
+
+    if (gcry_control(GCRYCTL_INITIALIZATION_FINISHED_P)) {
+
+        // Handle the hash algorithm
+        static auto* const ALGO = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:hash_algorithm"));
+        static auto const  algn = std::string(*ALGO);
+        m_iAlgo                 = gcry_md_map_name(algn.c_str());
+        m_iDigestLen            = gcry_md_get_algo_dlen(m_iAlgo);
+        if (m_iAlgo) {
+            static auto const err = gcry_err_code(gcry_md_test_algo(m_iAlgo));
+            if (err == GPG_ERR_NO_ERROR) {
+
+                // Handle the salt
+                static auto* const SALT = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:hash_salt"));
+                m_sSalt                 = std::string(*SALT);
```

Comment:
explicit cast redundant

Reply:
Without it, it's a `void*` which is not accepted by `std::string()`.